src/combinatorics/colex.lean

```diff
@@ -0,0 +1,293 @@
+/-- The colex ordering on finsets. -/
+def colex.le [has_lt α] (A B : finset α) : Prop := A.to_colex ≤ B.to_colex
+
+infix ` <ᶜ `:50 := colex.lt
+infix ` ≤ᶜ `:50 := colex.le
+
+lemma colex.lt_def [has_lt α] (A B : finset α) :
+  A <ᶜ B ↔ ∃ k, (∀ {x}, k < x → (x ∈ A ↔ x ∈ B)) ∧ k ∉ A ∧ k ∈ B :=
+iff.rfl
+lemma colex.le_def [has_lt α] (A B : finset α) :
+  A ≤ᶜ B ↔ A <ᶜ B ∨ A = B :=
+iff.rfl
+
+namespace colex
+
+/-- Strictly monotone functions preserve the colex ordering. -/
+lemma hom {β : Type*} [linear_order α] [decidable_eq β] [preorder β]
+  {f : α → β} (h₁ : strict_mono f) (A B : finset α) :
+  A.image f <ᶜ B.image f ↔ A <ᶜ B :=
+begin
+  simp only [colex.lt_def, not_exists, mem_image, exists_prop, not_and],
+  split,
+  { rintro ⟨k, z, q, k', _, rfl⟩,
+    exact ⟨k', λ x hx, by simpa [h₁.injective] using z (h₁ hx), λ t, q _ t rfl, ‹k' ∈ B›⟩ },
+  rintro ⟨k, z, ka, _⟩,
+  refine ⟨f k, λ x hx, _, _, k, ‹k ∈ B›, rfl⟩,
+  { split,
+    any_goals {
+      rintro ⟨x', hx', rfl⟩,
+      refine ⟨x', _, rfl⟩,
+      rwa ← z _ <|> rwa z _,
+      rwa strict_mono.lt_iff_lt h₁ at hx } },
+  { simp only [h₁.injective, function.injective.eq_iff],
+    exact λ x hx, ne_of_mem_of_not_mem hx ka }
+end
+
+/-- A special case of `colex_hom` which is sometimes useful. -/
+lemma colex_hom_fin {n : ℕ} (A B : finset (fin n)) :
+  A.image subtype.val <ᶜ B.image subtype.val ↔ A <ᶜ B :=
+colex.hom (λ x y k, k) _ _
+
+instance [has_lt α] : is_irrefl (finset.colex α) (<) :=
+⟨λ A h, exists.elim h (λ _ ⟨_,a,b⟩, a b)⟩
+
+@[trans]
+lemma lt_trans [linear_order α] {a b c : finset.colex α} :
+  a < b → b < c → a < c :=
+begin
+  rintros ⟨k₁, k₁z, notinA, inB⟩ ⟨k₂, k₂z, notinB, inC⟩,
+  cases lt_or_gt_of_ne (ne_of_mem_of_not_mem inB notinB),
+  { refine ⟨k₂, _, by rwa k₁z h, inC⟩,
+    intros x hx,
+    rw ← k₂z hx,
+    apply k₁z (trans h hx) },
+  { refine ⟨k₁, _, notinA, by rwa ← k₂z h⟩,
+    intros x hx,
+    rw k₁z hx,
+    apply k₂z (trans h hx) }
+end
+
+@[trans]
+lemma le_trans [linear_order α] (a b c : finset.colex α) :
+  a ≤ b → b ≤ c → a ≤ c :=
+λ AB BC, AB.elim (λ k, BC.elim (λ t, or.inl (lt_trans k t)) (λ t, t ▸ AB)) (λ k, k.symm ▸ BC)
+
+instance [linear_order α] : is_trans (finset.colex α) (<) := ⟨λ _ _ _, colex.lt_trans⟩
+instance [linear_order α] : is_asymm (finset.colex α) (<) := by apply_instance
+
+instance [linear_order α] : is_strict_order (finset.colex α) (<) := {}
+
+lemma lt_trichotomy [linear_order α] (A B : finset.colex α) :
+  A < B ∨ A = B ∨ B < A :=
+begin
+  classical,
+  by_cases (A = B),
+  { right,
+    left,
+    assumption  },
+  rcases (exists_max_image (A \ B ∪ B \ A) id _) with ⟨k, hk, z⟩,
+  { simp only [mem_union, mem_sdiff] at hk,
+    cases hk,
+    { right,
+      right,
+      refine ⟨k, λ t th, _, hk.2, hk.1⟩,
+      specialize z t,
+      by_contra,
+      simp only [mem_union, mem_sdiff, id.def] at z,
+      rw [not_iff, iff_iff_and_or_not_and_not, not_not, and_comm] at a,
```

Comment:
Looks like the error is here: I think it's from the `by_contra` introducing a variable called `a` in an old version but not any more: presumably changing line 163 to `by_contra a` fixes it?